docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,177 @@
+
+```protobuf
+
+// Capabilities of the driver
+message DriverCapabilities {
+	string supported_read_write_modes = 1;
+	enum AttachmentPolicy {
+	    NONE = 0;  // Driver does not support attach/detach.
+	    LOCAL = 1; // Driver supports local volume attachments from kubelet.
+	    REMOTE = 2; // Driver supports remote volume attachments to a node from controller-manager.
+	}
+	AttachmentPolicy supported_attachment_policy = 2;
+	bool supports_dynamic_provisioning = 3;
+	bool supports_custom_mount = 4;
+	bool supports_selinux = 5;
+	bool supports_ownership_management = 6;
+	bool supports_metrics = 7;
+}
+
+message ProbeDriverRequest {}
+
+message ProbeDriverResponse {
+    DriverCapabilities capabilities = 1;
+    repeated string options = 2;
+}
+
+service FlexVolumeService {
+    // Probe the driver capabilities.
+    rpc ProbeDriver(ProbeDriverRequest) returns (ProbeDriverResponse) {}
+    // Create a volume.
+    rpc Create(CreateRequest) returns (CreateResponse) {}
+    // Delete a volume.
+    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+    // Attach the volume.
+    rpc Attach(AttachRequest) returns (AttachResponse) {}
+    // Detach the volume.
+    rpc Detach(DetachRequest) returns (DetachResponse) {}
+    // Mount the volume.
+    rpc Mount(MountRequest) returns (MountResponse) {}
+    // unmount the volume.
+    rpc Unmount(UnmountRequest) returns (UnmountResponse) {}
+    // Query the metrics.
+    rpc GetMetrics(MetricsRequest) returns (MetricsResponse) {}
+}
+
+// Volume source.
+message Spec {
+    string name = 1; // Name of the volume.
+    string fstype = 2; // Requested file system type.
+    string read_write_mode = 3;
+    map<string, string> secrets = 4; // Secrets required for the driver to talk to its own controller.
+    map<string, string> options = 5; // Extra options passed to the driver.
+}
+
+message CreateRequest{
+    Spec spec = 1; // Volume specification.
+}
+
+message CreateResponse{}
+
+message DeleteRequest{
+    string name = 1; // Name of the volume.
+}
+
+message DeleteResponse{}
+
+message AttachRequest {
+    string name = 1; // Name of the volume.
+    string host = 2; // Name of the host to attach volume to.
+    bool sync = 3; // Wait for attach to finish.
+    map<string, string> secrets = 4; // Secrets required for the driver to talk to its own controller.
+    map<string, string> options = 5; // Extra options passed to the driver.
+}
+
+message AttachResponse{
+    string device_path = 1; // Path to the device where the volume is attached.
+}
+
+message DetachRequest {
+    string name = 1; // Name of the volume.
+    string host = 2; // Name of the host to detach volume from.
+    bool sync = 3; // Wait for detach to finish.
+    map<string, string> secrets = 4; // Extra options passed to the driver.
+}
+
+message DetachResponse {}
+
+message MountRequest {
+    Spec spec = 1; // Volume specifciation.
+    string mount_path = 2; // Recommended path for the driver to mount the volume.
```

Comment:
Using a different path that the suggested one won't work as kubelet won't know what devices to unmount when it is restarted. It needs the volumes mounted to a specific directory so it can keep track of them.